### PR TITLE
Modernize for v2: TypeScript, modern SDKs, promise API

### LINKED ISSUE #37: ERR_OUT_OF_RANGE error while running the library
Hello @jpizarrom @bendrucker

```
internal/validators.js:93
      throw new ERR_OUT_OF_RANGE(name, `>= ${min} && <= ${max}`, value);
      ^

RangeError [ERR_OUT_OF_RANGE]: The value of "length" is out of range. It must be >= 0 && <= 2147483647. Received 2598337812
    at Function.concat (buffer.js:561:5)
    at Request.<anonymous> (/usr/lib/node_modules/azure-blob-to-s3/node_modules/request/request.js:1126:30)
    at Request.emit (events.js:314:20)
    at Request.EventEmitter.emit (domain.js:483:12)
    at IncomingMessage.<anonymous> (/usr/lib/node_modules/azure-blob-to-s3/node_modules/request/request.js:1076:12)
    at Object.onceWrapper (events.js:420:28)
    at IncomingMessage.emit (events.js:326:22)
    at IncomingMessage.EventEmitter.emit (domain.js:483:12)
    at endReadableNT (_stream_readable.js:1241:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  code: 'ERR_OUT_OF_RANGE'
}
```

I'm getting this error while using the library 
I'm using the CLI way to run this.

```
node --version
v12.20.2

npm --version
6.14.11
```

### PR DESCRIPTION
**Pull request by bendrucker**
Rewrites the package in TypeScript as ESM-only for v2.0.0, replacing the EOL `azure-storage` and `aws-sdk` v2 SDKs with `@azure/storage-blob` and `@aws-sdk/client-s3` + `@aws-sdk/lib-storage`. `copy()` now returns a `Promise<CopySummary>` and reports progress through an `onProgress` callback instead of returning a stream. Closes #37, an error in the v1 stream stack that no longer exists in v2.

## Changes

- Delegated authentication to the cloud SDKs: Azure uses `DefaultAzureCredential` against the storage account URL (`azure.account` replaces `azure.connection`), and AWS uses the default credential provider chain. The `accessKeyId`/`secretAccessKey` options and CLI flags are gone, so secrets never pass through argv. Anything custom comes in as a user-constructed client through `azure.client` / `aws.client`; `copy()` only destroys clients it creates.
- Absorbed `azure-blob-list-stream` into a `listBlobsFlat().byPage()` loop. `page` progress events expose the continuation token so runs can resume.
- Replaced the stream glue (`through2`, `throttled-transform-stream`, `retry-stream-proxy`) with a bounded promise pool. `concurrency` now caps simultaneous transfers (default 30, `0` for unlimited); v1 throttled new transfers per second, defaulting to 100. I lowered the default because lib-storage already uploads 4 parts per file concurrently against a 50-socket pool, so large files saturate well before 30, while small latency-bound files still get enough parallelism. 100 simultaneous large-file uploads could also buffer ~2 GB of parts. Mid-stream download retry moves to the Azure SDK's `maxRetryRequests` option, which defaults to 0 and would otherwise silently drop v1's retry behavior.
- Errors fail fast: the first transfer error rejects `copy()` after in-flight transfers settle.
- Dropped `bole`, `ora`, and `queue`. The library no longer logs; the CLI prints ndjson via `console.log`, and `--log-level debug` adds per-file skip/upload events. The previously hidden `--azure-token` and `--log-level` flags are now documented.
- Publishes built JS plus type declarations from `dist/` only, with an `exports` map. Requires Node >= 22.18 so tests run through native type stripping (`node --test src/*.test.ts`).
- Toolchain: `tsc` for build and typecheck, neostandard (ts) for lint, Node 22 + 24 CI matrix with `npm ci`. I un-ignored `package-lock.json` since `npm ci` requires it. CI smoke-tests the packed tarball (`npm pack`, global install, `azure-s3 --help`), so the bin path lives only in `package.json`.

I kept v1's skip semantics (matching key and `Content-Length` on S3) and fixed two edge cases in the port: a blob with unknown `contentLength` always uploads instead of matching a 0-byte S3 object, and an empty `--aws-prefix` no longer produces keys with a leading slash.

## Testing

Replaced the tape/proxyquire suite with `node:test`: S3 mocked via `aws-sdk-client-mock` (covers `lib-storage`'s `Upload`), Azure via a structural `BlobContainer` fake injected through `azure.client`. Added coverage for prefixing, resume tokens, fail-fast head errors, the concurrency bound, `maxRetryRequests`, and both edge cases above.

## References

Next step after this lands: deprecate `azure-blob-list-stream` on npm and archive the repo, pointing users at `listBlobsFlat().byPage()`.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -7,17 +7,21 @@ name: tests
...
+        uses: actions/setup-node@v4
+        with:
+          node-version: ${{ matrix.node-version }}
+          cache: npm
+      - run: npm ci
+      - run: npm run lint
+      - run: npm run typecheck
+      - run: npm test
+      - run: npm run build
+      - run: node dist/cli.js --help
```
Comment (bendrucker) [inline .github/workflows/test.yml]: why are we pointing at the CLI directly by `dist/` path here?

File: src/cli.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  Options:
+    --concurrency            Maximum number of simultaneous transfers (default: 100)
+    --azure-connection       Azure Blob Storage connection string
+    --azure-container        Azure Blob Storage container name
+    --azure-token            Continuation token from a previous run's page output
+    --aws-bucket             AWS S3 bucket name
+    --aws-prefix             (Optional) A string between an Amazon S3 bucket name and an object name, for example: bucket/prefix1/file
+    --aws-region             AWS region for the bucket
+    --aws-access-key-id      AWS IAM access key ID
```
Comment (bendrucker) [inline src/cli.ts]: Since this is a major revision, we should go ahead and support proper AWS authentication. We should delegate authentication to the SDK and remove these flags. We certainly shouldn't be encouraging use of secrets in flags.

Same for Azure. We should delegate authentication to relevant packages and not try to implement it ourselves, and certainly not encourage or force users to use static secrets.

File: .github/workflows/test.yml (hunk 1 above)
Comment (bendrucker) [reply .github/workflows/test.yml]: This is the post-build smoke test. `bin` points at `dist/cli.js` and `npm pack` ships only `dist/`, so running the compiled entry verifies the published artifact, including the rewritten `./index.js` import, rather than the TS sources. It replaces v1's `node cli.js --help` check.

File: src/cli.ts (hunk 2 above)
Comment (bendrucker) [reply src/cli.ts]: Done in 0bb5514. Azure authenticates with `DefaultAzureCredential` against the storage account URL, so `--azure-account` replaces `--azure-connection`. AWS uses the SDK's default provider chain, and the static credential flags are gone.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+  }
+  return new ContainerClient(
+    `https://${account}.blob.core.windows.net/${container}`,
+    new DefaultAzureCredential()
+  )
+}
+
+export async function copy (options: CopyOptions): Promise<CopySummary> {
+  const concurrency = options.concurrency ?? 100
+
```
[the remark this answers is not in the document]
Comment (bendrucker) [reply src/index.ts:78]: Fixed in 98fb28f. `copy()` now throws a `RangeError` unless `concurrency` is a positive integer, with a test covering `concurrency: 0`.

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+  }
+
+  async function exists (key: string, size: number | undefined): Promise<boolean> {
+    try {
+      const head = await s3.send(new HeadObjectCommand({ Bucket: options.aws.bucket, Key: key }))
+      return size != null && head.ContentLength === size
+    } catch (error) {
+      if (error instanceof NotFound) return false
+      throw error
+    }
```
[the remark this answers is not in the document]
Comment (bendrucker) [reply src/index.ts:169]: Fixed in 98fb28f. `exists()` now matches on `error.name === 'NotFound'` instead of `instanceof`, so 404s from a user-supplied client built against another installed copy of `@aws-sdk/client-s3` are treated as missing. Added a test simulating the cross-copy case.
